lib/LanguageServerRename/Model/NodeUtils.php

```diff
@@ -0,0 +1,180 @@
+    {
+        return new Range(
+            PositionConverter::intByteOffsetToPosition($token->getStartPosition(), $document),
+            PositionConverter::intByteOffsetToPosition($token->getEndPosition(), $document)
+        );
+    }
+    
+    public function getNodeNameToken(Node $node, ?string $name = null): ?Token // NOSONAR
+    {
+        if ($node instanceof ClassDeclaration) {
+            return $node->name;
+        }
+        if ($node instanceof MethodDeclaration) {
+            return $node->name;
+        }
+        
+        if ($node instanceof InterfaceDeclaration) {
+            return $node->name;
+        }
+        
+        if ($node instanceof TraitDeclaration) {
+            return $node->name;
+        }
+        
+        if ($node instanceof QualifiedName && ($nameToken = $node->getLastNamePart()) !== null) {
+            return $nameToken;
+        }
+        
+        if ($node instanceof ScopedPropertyAccessExpression && $node->memberName instanceof Token) {
+            return $node->memberName;
+        }
+        
+        if ($node instanceof Variable && $node->name instanceof Token && $node->getFirstAncestor(PropertyDeclaration::class)) {
+            return $node->name;
+        }
+        
+        if ($node instanceof ConstElement) {
+            return $node->name;
+        }
+        
+        if ($node instanceof Parameter) {
+            return $node->variableName;
+        }
+        
+        if ($node instanceof Variable) {
+            while ($node->name instanceof Variable) {
+                $node = $node->name;
+            }
+            if ($node->name instanceof Token) {
+                return $node->name;
+            }
+            return null;
+        }
+        
+        if ($node instanceof MemberAccessExpression) {
+            return $node->memberName;
+        }
+        
+        if ($node instanceof ClassConstDeclaration) {
+            if (!empty($name)) {
+                foreach ($node->constElements->getElements() as $element) {
+                    if ($element instanceof ConstElement && $element->getName() == $name) {
+                        return $this->getNodeNameToken($element);
+                    }
+                }
+            }
+            return null;
+        }
+        
+        if ($node instanceof PropertyDeclaration) {
+            if (!empty($name)) {
+                foreach ($node->propertyElements->getElements() as $nodeOrToken) {
+                    /** @var Node|Token $nodeOrToken */
+                    if ($nodeOrToken instanceof Variable
+                        && $nodeOrToken->name instanceof Token
+                        && $nodeOrToken->getName() == $name
+                    ) {
+                        return $nodeOrToken->name;
+                    }
+                }
+            }
+            return null;
+        }
+
+        if ($node instanceof ForeachStatement && !empty($name)) {
```

Comment:
when would somebody rename a foreach statement?

Reply:
`foreach($variable1 as $variable => $variable3) {`
These are all variables and are subject to renaming

Reply:
Not to mention the case:
```php
$a = new class($var1) { 
    public function __construct($arg1) { .. }
}
```
which I don't think I have covered.

Reply:
but `foreach` isn't something you can rename? the variables in the exression are, but if you target them you won't get a foreach, you would, i think, get the `Variable`.

Reply:
If I am not mistaken references are returned in such a location that you get the foreach statement rather than the individual variables.

Reply:
I'm not sure why that would be the case? I would expect the reference to be `foreach (<>$foobar<> as $barfoo) {`

Reply:
So the locations are:
#### Array
```php
foreach(<>$references as $key=>$value)
```
#### Value
```php
<>foreach($references as $key=>$value)
```

#### Key
```php
<>foreach($references as $key=>$value)
```